src/Compilers/CSharp/Portable/FlowAnalysis/DataFlowPass.LocalFunctions.cs

```diff
@@ -0,0 +1,301 @@
+{
+    internal partial class DataFlowPass
+    {
+        private class LocalFuncInfo
+        {
+            public LocalFuncInfo(BoundLocalFunctionStatement boundNode)
+            {
+                BoundNode = boundNode;
+            }
+
+            public struct UsedVariable : IEqualityComparer<UsedVariable>
+            {
+                public bool Written { get; }
+                public Symbol CapturedSymbol { get; }
+                public Symbol SymbolOpt { get; }
+                public int? SlotOpt { get; }
+
+                public UsedVariable(bool written,
+                                    Symbol capturedSymbol,
+                                    Symbol symbol = null,
+                                    int? slot = null)
+                {
+                    Written = written;
+                    CapturedSymbol = capturedSymbol;
+                    SymbolOpt = symbol;
+                    SlotOpt = slot;
+                }
+
+                public bool Equals(UsedVariable left, UsedVariable right) =>
+                    left.Written == right.Written &&
+                    left.CapturedSymbol == right.CapturedSymbol &&
+                    left.SymbolOpt == right.SymbolOpt &&
+                    left.SlotOpt == right.SlotOpt;
+
+                public int GetHashCode(UsedVariable usedVar) =>
+                    Hash.Combine(usedVar.Written,
+                    Hash.Combine(usedVar.CapturedSymbol.GetHashCode(),
+                    Hash.Combine(usedVar.SymbolOpt?.GetHashCode() ?? 0,
+                                 usedVar.SlotOpt?.GetHashCode() ?? 0)));
+            }
+
+            public OrderedSet<UsedVariable> UsedVariables { get; } =
+                new OrderedSet<UsedVariable>();
+
+            private PooledHashSet<LocalFunctionSymbol> _usedLocalFuncs;
+            public PooledHashSet<LocalFunctionSymbol> UsedLocalFunctions
+            {
+                get
+                {
+                    if (_usedLocalFuncs == null)
+                    {
+                        _usedLocalFuncs = PooledHashSet<LocalFunctionSymbol>.GetInstance();
+                    }
+
+                    return _usedLocalFuncs;
+                }
+            }
+
+            /// <summary>
+            /// Indicates that the set of "used" variables changed during the
+            /// last data flow pass. If the set changed, we must re-run the
+            /// analysis until we have reached a fixed-point.
+            /// </summary>
+            public bool IsDirty { get; set; } = false;
+
+            public BoundLocalFunctionStatement BoundNode { get; }
+
+            public void Free()
+            {
+                _usedLocalFuncs?.Free();
+            }
+        }
+
+        // Save the list of reads, writes, and diagnostics for all local funcs
+        private SmallDictionary<LocalFunctionSymbol, LocalFuncInfo> _localFuncResults;
+        private SmallDictionary<LocalFunctionSymbol, LocalFuncInfo> LocalFuncResults
```

Comment:
Please move this to the top of this partial class part. We prefer to keep all field declarations at the top.